File: ios/Capacitor/Capacitor/Plugins/Toast.swift
import Foundation

@objc(CAPToastPlugin)
public class CAPToastPlugin: CAPPlugin {

  @objc func show(_ call: CAPPluginCall) {
    guard let text = call.get("text", String.self) else {
      call.error("text must be provided and must be a string.")
      return
    }
    let durationType = call.get("duration", String.self, "short")!
    let duration = durationType == "long" ? 3500 : 2000
    let position = call.get("position", String.self, "bottom")!

    guard let viewController = bridge?.viewController else {
      call.error("Unable to display toast!")
      return
    }
    showToast(in: viewController, text: text, duration: duration, position: position, completion: {(_) in
      call.success()
    })
  }

  public func showToast(in viewController: UIViewController, text: String, duration: Int = 2000, position: String = "bottom", completion: ((Bool) -> Void)? = nil) {
    DispatchQueue.main.async {
      let maxSizeTitle: CGSize = CGSize(width: viewController.view.bounds.size.width-32, height: viewController.view.bounds.size.height)

      let label = UILabel()
      label.backgroundColor = UIColor.black.withAlphaComponent(0.6)
      label.textColor = UIColor.white
      label.textAlignment = .center
      label.text = text
      label.alpha = 0
      label.layer.cornerRadius = 18
      label.clipsToBounds  =  true
      label.lineBreakMode = .byWordWrapping
      label.numberOfLines = 0

      var expectedSizeTitle: CGSize = label.sizeThatFits(maxSizeTitle)
      // UILabel can return a size larger than the max size when the number of lines is 1
      let minWidth = min(maxSizeTitle.width, expectedSizeTitle.width)
      let minHeight = min(maxSizeTitle.height, expectedSizeTitle.height)
      expectedSizeTitle = CGSize(width: minWidth, height: minHeight)

      let height = expectedSizeTitle.height+32
      let yCoordinate: CGFloat
      if position == "top" {
        yCoordinate = 40
      } else if position == "center" {
        yCoordinate = (viewController.view.bounds.size.height/2) - (height/2)
      } else {
        yCoordinate = viewController.view.bounds.size.height - height - (height/2)
      }

      label.frame = CGRect(
        x: ((viewController.view.bounds.size.width)/2) - ((expectedSizeTitle.width+32)/2),
        y: yCoordinate,
        width: expectedSizeTitle.width+32,
        height: height)

      label.padding = UIEdgeInsets(top: 8, left: 8, bottom: 8, right: 8)

      viewController.view.addSubview(label)

      UIView.animateKeyframes(withDuration: 0.3, delay: 0, animations: {
        label.alpha = 1.0
      }, completion: {(isCompleted) in

        UIView.animate(withDuration: 0.3, delay: (Double(duration) / 1000), options: .curveEaseOut, animations: {
          label.alpha = 0.0
        }, completion: {(isCompleted) in
          label.removeFromSuperview()

          if (completion) != nil {
            completion?(isCompleted)
          }
        })
      })
    }
  }
}

File: android/capacitor/src/main/java/com/getcapacitor/plugin/Toast.java
package com.getcapacitor.plugin;

import android.view.Gravity;
import com.getcapacitor.NativePlugin;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;

@NativePlugin()
public class Toast extends Plugin {

  private static final int GRAVITY_TOP = Gravity.TOP|Gravity.CENTER_HORIZONTAL;
  private static final int GRAVITY_CENTER = Gravity.CENTER_VERTICAL|Gravity.CENTER_HORIZONTAL;

  @PluginMethod()
  public void show(PluginCall call) {
    CharSequence text = call.getString("text");
    if(text == null) {
      call.error("Must provide text");
      return;
    }

    String durationType = call.getString("durationType", "short");

    int duration = android.widget.Toast.LENGTH_SHORT;
    if("long".equals(durationType)) {
      duration = android.widget.Toast.LENGTH_LONG;
    }

    android.widget.Toast toast = android.widget.Toast.makeText(getContext(), text, duration);

    String position = call.getString("position", "bottom");
    if("top".equals(position)) {
      toast.setGravity(GRAVITY_TOP, 0, 40);
    } else if("center".equals(position)) {
      toast.setGravity(GRAVITY_CENTER, 0, 0);
    }

    toast.show();
    call.success();
  }
}
